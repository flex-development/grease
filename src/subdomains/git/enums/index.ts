/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
