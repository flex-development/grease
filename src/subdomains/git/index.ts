/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './options'
export * from './providers'
export * from './queries'
export type * from './types'
