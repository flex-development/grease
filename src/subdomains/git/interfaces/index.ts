/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
