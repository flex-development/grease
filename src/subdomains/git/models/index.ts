/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
