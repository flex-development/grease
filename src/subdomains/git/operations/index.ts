/**
 * @file Entry Point - Operations
 * @module grease/git/operations
 */

export { default as TagOperationHandler } from './tag.handler'
export { default as TagOperation, type TagOperationDTO } from './tag.operation'
