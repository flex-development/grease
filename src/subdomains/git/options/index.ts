/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitOptions } from './git.options'
