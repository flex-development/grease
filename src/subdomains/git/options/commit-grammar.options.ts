/**
 * @file Options - CommitGrammarOptions
 * @module grease/git/options/CommitGrammarOptions
 */

import { RepositoryProvider } from '#src/git/enums'
import {
  get,
  isString,
  select,
  sort,
  unique,
  type Partial
} from '@flex-development/tutils'
import {
  IsArray,
  IsEnum,
  IsString,
  MinLength
} from 'class-validator'

/**
 * Commit parser grammar options.
 *
 * @class
 */
class CommitGrammarOptions {
  /**
   * Reference action keywords.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} actions
   */
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  public actions: string[]

  /**
   * Issue reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} issues
   */
  @IsArray()
  @IsString({ each: true })
  public issues: string[]

  /**
   * Pull request reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} pr
   */
  @IsArray()
  @IsString({ each: true })
  public pr: string[]

  /**
   * Repository provider name.
   *
   * @default RepositoryProvider.GITHUB
   *
   * @public
   * @instance
   * @member {RepositoryProvider} provider
   */
  @IsEnum(RepositoryProvider)
  public provider: RepositoryProvider

  /**
   * Tag prefix.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} tagprefix
   */
  @IsString()
  public tagprefix: string

  /**
   * Create a new options object.
   *
   * @param {Partial<CommitGrammarOptions>?} [opts] - Commit grammar options
   */
  constructor(opts?: Partial<CommitGrammarOptions>) {
    this.actions = select(get(opts, 'actions', []), isString)
    this.issues = select(get(opts, 'issues', []), isString)
    this.pr = select(get(opts, 'pr', []), isString)
    this.provider = get(opts, 'provider', RepositoryProvider.GITHUB)
    this.tagprefix = get(opts, 'tagprefix', '')

    // add common provider actions and prefixes
    this.actions.push('close', 'closed', 'closes', 'fix', 'fixed')
    this.issues.push('#')
    this.provider !== RepositoryProvider.GITLAB && this.pr.push('#')

    // add actions and prefixes based on provider
    switch (this.provider) {
      case RepositoryProvider.BITBUCKET:
        this.actions.push(
          'closing',
          'fixes',
          'fixing',
          'resolve',
          'resolved',
          'resolves',
          'resolving'
        )
        this.issues.unshift('JRA-', 'JIRA-')
        break
      case RepositoryProvider.GITLAB:
        this.actions.push('closing', 'fixes', 'fixing')
        this.pr.push('!')
        break
      default:
        this.actions.push('fixes', 'resolve', 'resolved', 'resolves')
        this.issues.push('GH-')
        break
    }

    // ensure actions and prefixes are unique
    this.actions = sort(unique(this.actions), (a, b) => b.length - a.length)
    this.issues = unique(this.issues)
    this.pr = unique(this.pr)
  }
}

export default CommitGrammarOptions
