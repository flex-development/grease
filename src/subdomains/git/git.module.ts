/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { CqrsModule } from '@nestjs/cqrs'
import { TagOperationHandler } from './operations'
import { GitService } from './providers'
import { CommitQueryHandler, TagQueryHandler } from './queries'

/**
 * Git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  imports: [CqrsModule],
  providers: [
    CommitQueryHandler,
    GitService,
    LoggerService,
    TagOperationHandler,
    TagQueryHandler,
    ValidationService
  ]
})
class GitModule {}

export default GitModule
