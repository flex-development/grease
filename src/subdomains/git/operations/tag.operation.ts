/**
 * @file Operations - TagOperation
 * @module grease/git/operations/TagOperation
 */

import { IsBoolable } from '#src/decorators'
import { GitOptions } from '#src/git/options'
import {
  DOT,
  defaults,
  ifelse,
  isFalsy,
  isUndefined,
  trim,
  type Assign
} from '@flex-development/tutils'
import { IsBoolean, IsNotIn, IsString, MinLength } from 'class-validator'

/**
 * Tag operation payload transfer object.
 */
type TagOperationDTO = Assign<
  Partial<TagOperation>,
  Pick<TagOperation, 'tag'>
>

/**
 * Git tag operation.
 *
 * @see {@linkcode GitOptions}
 * @see https://git-scm.com/docs/git-tag
 *
 * @class
 * @extends {GitOptions}
 */
class TagOperation extends GitOptions {
  /**
   * Replace an existing tag instead of failing.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} force
   */
  @IsBoolean()
  public force: boolean

  /**
   * Tag message.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} message
   */
  @IsString()
  public message: string

  /**
   * Object the new tag will refer to.
   *
   * @default 'HEAD'
   *
   * @public
   * @instance
   * @member {string} object
   */
  @IsString()
  @MinLength(1)
  public object: string

  /**
   * Push tag to {@linkcode remote} after successful creation.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} push
   */
  @IsBoolean()
  public push: boolean

  /**
   * Destination of the {@linkcode push} operation.
   *
   * This value may be the name of a remote, or a git-compliant URL.
   *
   * @see https://git-scm.com/docs/git-push#REMOTES
   * @see https://git-scm.com/docs/git-push#URLS
   *
   * @default 'origin'
   *
   * @public
   * @instance
   * @member {string} remote
   */
  @IsString()
  @MinLength(1)
  public remote: string

  /**
   * Create a GPG-signed tag, using the default e-mail address' key if `true`,
   * or the given key if a string.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean | string} sign
   */
  @IsString()
  @MinLength(1)
  @IsBoolable()
  public sign: boolean | string

  /**
   * Name of tag to create.
   *
   * The name must pass all checks defined by [`git-check-ref-format`][1]. Some
   * checks may restrict the characters allowed in a tag name.
   *
   * [1]: https://git-scm.com/docs/git-check-ref-format
   *
   * @public
   * @instance
   * @member {string} tag
   */
  @IsString()
  @MinLength(1)
  @IsNotIn(['@', '?', '['])
  public tag: string

  /**
   * Verify GPG signature of the new tag.
   *
   * @default !isFalsy(sign)
   *
   * @public
   * @instance
   * @member {boolean} verify
   */
  @IsBoolean()
  public verify: boolean

  /**
   * Create a new tag operation.
   *
   * @see {@linkcode TagOperationDTO}
   *
   * @param {TagOperationDTO} payload - Operation payload
   */
  constructor(payload: TagOperationDTO) {
    super(payload)

    const {
      force,
      message,
      object,
      push,
      remote,
      sign,
      tag,
      verify
    } = defaults(payload, {
      force: false,
      message: '',
      object: 'HEAD',
      push: false,
      remote: 'origin',
      sign: false,
      tag: '',
      verify: false
    })

    this.force = force
    this.message = message
    this.object = object
    this.push = push
    this.remote = trim(remote)
    this.sign = JSON.parse(JSON.stringify(sign))
    this.tag = ifelse(trim(tag), this.tagprefix + tag, DOT)
    this.verify = verify

    isUndefined(payload.verify) && !isFalsy(this.sign) && (this.verify = true)
  }
}

export { TagOperation as default, type TagOperationDTO }
