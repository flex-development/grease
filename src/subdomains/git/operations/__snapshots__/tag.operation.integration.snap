// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:git/operations/TagOperation > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "force must be a boolean value",
    },
    "property": "force",
    "value": 1,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "message must be a string",
    },
    "property": "message",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "object must be a string",
      "minLength": "object must be longer than or equal to 1 characters",
    },
    "property": "object",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "push must be a boolean value",
    },
    "property": "push",
    "value": 0,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "remote must be longer than or equal to 1 characters",
    },
    "property": "remote",
    "value": "",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "sign must be a string",
      "minLength": "sign must be longer than or equal to 1 characters",
    },
    "property": "sign",
    "value": 1,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isNotIn": "tag should not be one of the following values: @, ?, [",
    },
    "property": "tag",
    "value": "?",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "verify must be a boolean value",
    },
    "property": "verify",
    "value": 1,
  },
]
`;
